### PR TITLE
Implement route for loading register page

### PR DESCRIPTION
**Pull request by haishengwu-okta**
## Description:

- Add route `/signin/register` and show registration form directly
- Show terminal error if registration is not supported.

## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Added unit tests?
- [ ] Added e2e tests
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:
Screen Shot 2021-03-31 at 1 45 21 PM
Screen Shot 2021-03-31 at 1 45 00 PM

### Issue:

- okta-380941

### CODE REVIEW

File: src/v2/controllers/RegistrationFormController.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    // in session storage
+    // - try to sign-in
+    // - during at middle of sign-in, change URL to /signin/register/
+    // user probably expects to see sign-up page
+    // actually widget will continue show sign-in flow (because of the saved state token)
+    if (this.options.appState.get('currentFormName') === FORMS.IDENTIFY) {
+      // 1. auto switch to porfile enroll (register) page
+      // basically mimic the flow that show identify page and click sign-up link.
+      if (this.options.appState.hasRemediationObject(FORMS.SELECT_ENROLL_PROFILE)) {
+        this.handleInvokeAction(FORMS.SELECT_ENROLL_PROFILE);
```
Comment (nikhilvenkatraman-okta) [inline src/v2/controllers/RegistrationFormController.js:28]: nice, saving users a click!

File: src/v2/controllers/RegistrationFormController.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      } else {
+        const messages = {
+          'type': 'array',
+          value: [
+            {
+              message: 'registration is not supported',
```
Comment (nikhilvenkatraman-okta) [inline src/v2/controllers/RegistrationFormController.js:34]: nit: this looks like a custom client side error ? let us make sure to add i18n key to keep it consistent with the rest of our API responses.

File: src/v2/controllers/RegistrationFormController.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          // OKTA-382410 so bad that has to leverage rawIdxState
+          rawIdxState: {
+            messages
+          },
+          context: {
+            messages
+          }
+        };
+        // this.options.appState.setIonResponse(resp);
+        this.options.appState.trigger('remediationSuccess', resp);
```
Comment (nikhilvenkatraman-okta) [inline src/v2/controllers/RegistrationFormController.js:51]: When we show them the error in the terminal state `registration is not supported` should we have a link for them to go back to the login page ?

File: src/v2/controllers/RegistrationFormController.js (hunk 2 above)
Comment (haishengwu-okta) [reply src/v2/controllers/RegistrationFormController.js:34]: Surely yes!

File: src/v2/controllers/RegistrationFormController.js (hunk 3 above)
Comment (haishengwu-okta) [reply src/v2/controllers/RegistrationFormController.js:51]: Good point!

File: test/testcafe/framework/page-objects/components/BaseFormObject.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import { Selector, ClientFunction } from 'testcafe';
 
 const TERMINAL_CONTENT = '.o-form-error-container .ion-messages-container';
-const FORM_INFOBOX_ERROR = '.okta-form-infobox-error';
+const FORM_INFOBOX_ERROR = '.o-form-error-container .infobox-error';
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/framework/page-objects/components/BaseFormObject.js:4]: nit: Consider updating these to use data-se attributes https://github.com/okta/okta-signin-widget/blob/14abeb1db905f124765a684cc2f0cc9526331a91/test/testcafe/framework/page-objects/ChallengeOktaVerifyPushPageObject.js#L4

File: test/testcafe/spec/ChallengeOktaVerifyFastPassView_spec.js [hunk 5]
```diff
@@ -47,7 +48,7 @@ const loopbackFallbackMock = RequestMock()
   .respond(identifyWithUserVerificationCustomURI);
 
 const identifyWithLaunchAuthenticatorHttpCustomUri = JSON.parse(JSON.stringify(identifyWithUserVerificationCustomURI));
-const mockHttpCustomUri = 'http://localhost:3000/launch-okta-verify';
+const mockHttpCustomUri = 'http://localhost:6512/launch-okta-verify';
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/spec/ChallengeOktaVerifyFastPassView_spec.js:51]: This change seems unrelated to profile enrollment and I didn't see anything in the description for this. Is this intentional ?

File: test/testcafe/spec/DeviceChallengePollView_spec.js [hunk 6]
```diff
@@ -102,7 +111,7 @@ const loopbackFallbackMock = RequestMock()
   .respond(identifyWithLaunchAuthenticator);
 
 const identifyWithLaunchAuthenticatorHttpCustomUri = JSON.parse(JSON.stringify(identifyWithLaunchAuthenticator));
-const mockHttpCustomUri = 'http://localhost:3000/launch-okta-verify';
+const mockHttpCustomUri = 'http://localhost:6512/launch-okta-verify';
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/spec/DeviceChallengePollView_spec.js:114]: same here

File: test/testcafe/spec/IdentifyRegistration_spec.js [hunk 7]
```diff
@@ -154,36 +152,86 @@ test.requestHooks(enrollProfileErrorMock)('should show email field validation er
...
+  await registrationPage.fillLastNameField('xyz');
+  await registrationPage.fillEmailField('foo@ex.com');
+  await registrationPage.clickRegisterButton();
+
+  // show successful terminal view and fires after render event
+  await t.expect(registrationPage.getTerminalContent()).eql(
+    'An activation email has been sent to john@gmail.com. Follow instructions in the email to finish creating your account'
+  );
+
+  const { log: log2 } = await t.getBrowserConsoleMessages();
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/spec/IdentifyRegistration_spec.js:172]: Can we use shared util `checkConsoleMessages` here ? https://github.com/okta/okta-signin-widget/blob/ec377eeb098304af04aeb93427b4538b7503f76c/test/testcafe/spec/Introspect_spec.js#L24

File: test/testcafe/spec/ChallengeOktaVerifyFastPassView_spec.js (hunk 5 above)
Comment (haishengwu-okta) [reply test/testcafe/spec/ChallengeOktaVerifyFastPassView_spec.js:51]: Yeah, due to change in `webpack.playground`, the playground would support any URL.
Before the change, `http://localhost:3000/launch-okta-verify` will respond 404 since the test case didn't care whether the URL works or not. But now the URL loads correctly and will generates additional logs to console hence test failed.

File: test/testcafe/spec/DeviceChallengePollView_spec.js (hunk 6 above)
Comment (haishengwu-okta) [reply test/testcafe/spec/DeviceChallengePollView_spec.js:114]: relied 👆

File: src/v2/WidgetRouter.js [hunk 8]
```diff
@@ -10,17 +10,21 @@
...
     '': 'defaultAuth',
+    'signin/register': 'renderRegister',
     '*wildcard': 'defaultAuth',
   },
 
   defaultAuth: function() {
     this.render(FormController);
   },
-
+  renderRegister() {
```
Comment (pradeepdewda-okta) [inline src/v2/WidgetRouter.js:27]: can you add a comment here, why we are doing this and why it is different then `defaultAuth` ?

File: src/v2/controllers/RegistrationFormController.js (hunk 3 above)
Comment (pradeepdewda-okta) [reply src/v2/controllers/RegistrationFormController.js:54]: should we even show this error message to enduser ? It will not make sense to them..  

I think user should not even see the page or link, if it is not enabled at the first place. Like pre-validation and avoid this state somehow.

File: src/v2/controllers/RegistrationFormController.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import FormController from './FormController';
+import { FORMS } from '../ion/RemediationConstants';
+
+export default FormController.extend({
```
Comment (pradeepdewda-okta) [inline src/v2/controllers/RegistrationFormController.js:16]: Add a comment that why we are adding a new `RegistrationFormController`

File: src/v2/controllers/RegistrationFormController.js (hunk 3 above)
Comment (haishengwu-okta) [reply src/v2/controllers/RegistrationFormController.js:54]: Assume you meant default login page where server side can valid and return 40x.
How we gonna avoid this for customer hosted page where we have no control of server but only the widget?

**Review by nikhilvenkatraman-okta — APPROVED** [review]

File: src/v2/controllers/RegistrationFormController.js (hunk 3 above)
Comment (pradeepdewda-okta) [reply src/v2/controllers/RegistrationFormController.js:62]: Instead of showing this error we can redirect user to some default page with warning like `Registration form is not enabled for this org`

and show them default login page.. 

I am inclining toward not showing errors (specially for those , for which enduser's will scratch their heads) like `Invalid state token provided`..

**Review by pradeepdewda-okta — CHANGES_REQUESTED** [review]
Instead of showing error, redirect to something default with warning on why last action was not succeeded.

File: src/v2/controllers/RegistrationFormController.js (hunk 3 above)
Comment (haishengwu-okta) [reply src/v2/controllers/RegistrationFormController.js:63]: Add another commit to customize the message and add `back to sign-in` link.

**Review by pradeepdewda-okta — APPROVED** [review]
thanks.. 💯

File: packages/@okta/i18n/src/properties/login.properties [hunk 10]
```diff
@@ -927,6 +927,7 @@ oie.selfservice.unlock_user.success.message = You can log in using your existing
 
 ## Terminal errors
 oie.session.expired = The session has expired.
+oie.registeration.is.not.enabled = Sign up is not enabled for this organization.
```
Comment (anipendakur-okta) [inline packages/@okta/i18n/src/properties/login.properties:930]: `nit`: typo - `registration` (vs `registeration`)

**Comment by kshestakov** [comment]
@anipendakur-okta @pradeepdewda-okta @nikhilvenkatraman-okta @haishengwu-okta Is this PR going to be merged???
